src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AccessController.cs

```diff
@@ -0,0 +1,554 @@
+                });
+            }
+
+            if (request.HasScope(OpenIdConnectConstants.Scopes.OfflineAccess) && !application.AllowRefreshTokenFlow)
+            {
+                return BadRequest(new OpenIdConnectResponse
+                {
+                    Error = OpenIdConnectConstants.Errors.InvalidRequest,
+                    ErrorDescription = T["Offline scope is not allowed for this OpenID Connect Application"]
+                });
+            }
+
+            if (request.IsPasswordGrantType())
+            {
+                if (!application.AllowPasswordFlow)
+                {
+                    return BadRequest(new OpenIdConnectResponse
+                    {
+                        Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                        ErrorDescription = T["Password Flow is not allowed for this OpenID Connect Application"]
+                    });
+                }
+
+                return await ExchangePasswordGrantType(request);
+            }
+
+            if (request.IsClientCredentialsGrantType())
+            {
+                if (!application.AllowClientCredentialsFlow)
+                {
+                    return BadRequest(new OpenIdConnectResponse
+                    {
+                        Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                        ErrorDescription = T["Client Credentials Flow is not allowed for this OpenID Connect Application"]
+                    });
+                }
+
+                return await ExchangeClientCredentialsGrantType(request);
+            }
+
+            if (request.IsAuthorizationCodeGrantType())
+            {
+                if (!application.AllowAuthorizationCodeFlow)
+                {
+                    return BadRequest(new OpenIdConnectResponse
+                    {
+                        Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                        ErrorDescription = T["Authorization Code Flow is not allowed for this OpenID Connect Application"]
+                    });
+                }
+
+                return await ExchangeAuthorizationCodeOrRefreshTokenGrantType(request);
+            }
+
+            if (request.IsRefreshTokenGrantType())
+            {
+                if (!application.AllowRefreshTokenFlow)
+                {
+                    return BadRequest(new OpenIdConnectResponse
+                    {
+                        Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                        ErrorDescription = T["Refresh Token Flow is not allowed for this OpenID Connect Application"]
+                    });
+                }
+
+                return await ExchangeAuthorizationCodeOrRefreshTokenGrantType(request);
+            }
+
+            throw new NotSupportedException("The specified grant type is not supported.");
```

Comment:
It is. If you reach this code path, this means you've enabled a flow in the OpenIddict options without adding the proper support in your controller code. This could be a `Debug.Fail()`, but this wouldn't make the compiler happy, so `NotSupportedException` it is.